### PR TITLE
Multiple curative states

### PR DESCRIPTION
**Pull request by bqth29**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?**

This PR allows multiple curative instants to be handled by the optimization.

**What is the new behavior (if this is a feature change)?**

Changes were also added to create `ContingencyScenario` objects to handle multiple curative instants and the CNEC-less states. Depending on the availability of ARAs/CRAs and auto/curative CNECs, different actions are to be taken following this truth table (for one curative instant only):

| ARA | Auto CNECs | CRA | Curative CNECSs | Action                                                            |
| --- | ---------- | --- | --------------- | ----------------------------------------------------------------- |
| ❌   | ❌          | ❌   | ❌               | Ignore                                                            |
| ❌   | ❌          | ❌   | ✅               | Add curative CNECs in preventive perimeter                        |
| ❌   | ❌          | ✅   | ❌               | Ignore                                                            |
| ❌   | ❌          | ✅   | ✅               | Create curative perimeter                                         |
| ❌   | ✅          | ❌   | ❌               | Add auto CNECs in preventive perimeter (duplicate with outage)    |
| ❌   | ✅          | ❌   | ✅               | Add auto and curative CNECs in preventive perimeter               |
| ❌   | ✅          | ✅   | ❌               | Add auto CNECs in preventive perimeter                            |
| ❌   | ✅          | ✅   | ✅               | Add auto CNECs in preventive perimeter, create curative perimeter |
| ✅   | ❌          | ❌   | ❌               | Ignore                                                            |
| ✅   | ❌          | ❌   | ✅               | Add auto perimeter (to apply SPS) and curative perimeter                      |
| ✅   | ❌          | ✅   | ❌               | Ignore                                                            |
| ✅   | ❌          | ✅   | ✅               | Create auto and curative perimeter                                         |
| ✅   | ✅          | ❌   | ❌               | Create auto perimeter                                             |
| ✅   | ✅          | ❌   | ✅               | Create auto and curative perimeters (but no CRAs available)       |
| ✅   | ✅          | ✅   | ❌               | Create auto perimeter                                             |
| ✅   | ✅          | ✅   | ✅               | Create auto and curative perimeters                               |

**Does this PR introduce a breaking change or deprecate an API?**
- [x] Yes
- [ ] No

**What changes might users need to make in their application due to this PR? (migration steps)**

The class `BasecaseScenario` was deleted and replaced by `Perimeter` that is mutualized for both preventive and curative perimeters.

### CODE REVIEW

File: ra-optimisation/search-tree-rao/src/test/java/com/powsybl/openrao/searchtreerao/castor/algorithm/CastorFullOptimizationTest.java [hunk 1]
```diff
@@ -663,4 +664,212 @@ void testOptimizationStepsExecutedAndLogsWhenFallbackOnFirstPrev() {
...
+
+        RaoInput raoInput = RaoInput.build(network, crac).build();
+        RaoParameters raoParameters = JsonRaoParameters.read(getClass().getResourceAsStream("/parameters/RaoParameters_2P_v2.json"));
+        raoParameters.getObjectiveFunctionParameters().setType(ObjectiveFunctionParameters.ObjectiveFunctionType.MAX_MIN_MARGIN_IN_AMPERE);
+
+        //co.apply(network, null);
+        //pstPrev.apply(network);
+        //naCur1.apply(network);
+        //pstCur.apply(network);
+        //LoadFlow.find("OpenLoadFlow").run(network, raoParameters.getLoadFlowAndSensitivityParameters().getSensitivityWithLoadFlowParameters().getLoadFlowParameters());
```
Comment (pet-mit) [inline ra-optimisation/search-tree-rao/src/test/java/com/powsybl/openrao/searchtreerao/castor/algorithm/CastorFullOptimizationTest.java:716]: remove

File: ra-optimisation/search-tree-rao/src/test/java/com/powsybl/openrao/searchtreerao/castor/algorithm/CastorFullOptimizationTest.java [hunk 2]
```diff
@@ -663,4 +664,212 @@ void testOptimizationStepsExecutedAndLogsWhenFallbackOnFirstPrev() {
...
+        assertEquals(1267.27, raoResult.getMargin(crac.getInstant("curative3"), cnec, Unit.AMPERE), 1.);
+
+        assertEquals(671.88, raoResult.getFunctionalCost(null), 1.);
+        assertEquals(429.22, raoResult.getFunctionalCost(crac.getInstant(InstantKind.PREVENTIVE)), 1.);
+        assertEquals(-20.30, raoResult.getFunctionalCost(crac.getInstant("curative1")), 1.); // TODO
+        assertEquals(-20.30, raoResult.getFunctionalCost(crac.getInstant("curative3")), 1.);
+    }
+
+    @Test
+    void testCurative2Instant() {
```
Comment (pet-mit) [inline ra-optimisation/search-tree-rao/src/test/java/com/powsybl/openrao/searchtreerao/castor/algorithm/CastorFullOptimizationTest.java:762]: change name

**Review by pet-mit — CHANGES_REQUESTED** [review]

File: ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/castor/algorithm/StateTree.java [hunk 3]
```diff
@@ -55,6 +58,136 @@ private void processOutageInstant(Contingency contingency, Crac crac) {
...
+            return;
+        }
+
+        boolean scenarioHasCurativeStates;
+        Instant firstCurativeInstantWithRemedialActions = instantsWithRemedialActions.isEmpty() ? null : instantsWithRemedialActions.stream().toList().get(0);
+        Instant lastCurativeInstantWithRemedialActions = instantsWithRemedialActions.isEmpty() ? null : instantsWithRemedialActions.stream().toList().get(instantsWithRemedialActions.size() - 1);
+
+        // get all the states the instants of which occur before the first curative instant with CRAs:
+        // - if ARAs are defined in the CRAC, create a curative perimeter using the first of these states (chronologically)
+        // - otherwise, add them all to the preventive perimeter
```
Comment (pet-mit) [inline ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/castor/algorithm/StateTree.java:105]: simplify and put in javadoc

**Review by pet-mit — CHANGES_REQUESTED** [review]

File: ra-optimisation/search-tree-rao/src/test/java/com/powsybl/openrao/searchtreerao/result/impl/PreventiveAndCurativesRaoResultImplTest.java [hunk 4]
```diff
@@ -1022,4 +1026,6 @@ void testAngleAndVoltageCnec() {
         exception = assertThrows(OpenRaoException.class, () -> output.getMargin(optimizedInstant, angleCnec, AMPERE));
         assertEquals("Angle cnecs are not computed in the rao", exception.getMessage());
     }
+
+    // TODO: add UT that checks the behaviour of findStateOptimizedFor
```
Comment (pet-mit) [inline ra-optimisation/search-tree-rao/src/test/java/com/powsybl/openrao/searchtreerao/result/impl/PreventiveAndCurativesRaoResultImplTest.java:1030]: do this

File: ra-optimisation/search-tree-rao/src/test/java/com/powsybl/openrao/searchtreerao/result/impl/PreventiveAndCurativesRaoResultImplTest.java [hunk 5]
```diff
@@ -140,6 +141,9 @@ public void setUp() {
         when(curativeState2.getInstant()).thenReturn(curativeInstant);
         when(curativeState2.getContingency()).thenReturn(Optional.of(contingency2));
         when(curativeState3.getInstant()).thenReturn(curativeInstant);
+        when(crac.getInstantBefore(curativeInstant)).thenReturn(autoInstant);
```
Comment (pet-mit) [inline ra-optimisation/search-tree-rao/src/test/java/com/powsybl/openrao/searchtreerao/result/impl/PreventiveAndCurativesRaoResultImplTest.java:144]: at this point, I think we may as well replace all this with an actual crac creation

**Review by pet-mit — CHANGES_REQUESTED** [review]

**Review by pet-mit — APPROVED** [review]

File: ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/castor/algorithm/StateTree.java [hunk 6]
```diff
@@ -59,36 +65,126 @@ private void processOutageInstant(Contingency contingency, Crac crac) {
      * If the state has AUTO and CURATIVE RAs, both states will be treated in a dedicated scenario.
      * If the AUTO has no RA but the CURATIVE has RAs, the AUTO will be optimized in basecase RAO and the CURATIVE in a dedicated scenario.
      * If neither AUTO nor CURATIVE states have RAs, they will be optimized in basecase RAO.
-     *
+     * <p>
      * If AUTO or CURATIVE state does not exist, it will not be optimized.
```
Comment (phiedw) [inline ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/castor/algorithm/StateTree.java:69]: Should we update this description?

**Pull request merged**